### PR TITLE
Memory reservation & metrics for cross join

### LINKED ISSUE #5162: Ignoring of memory-pool limits & OOM on large cartesian-product join
**Describe the bug**
There is an issue with possible OOM instead of `ResourcesExhausted`
Probably related to usage of unbounded channels (I believe it should be avoided actually)

**To Reproduce**

MRE to achieve ignoring of memory-pool with large Cartesian product:

CSV File example (250mb): [GDrive link](https://drive.google.com/file/d/1q_-p8BvvO2w-0IH7SyxvDIOYK44yQIKt/view?usp=share_link) - it's random file and column to join by has the same value for all records (so it's cartesian product)

Memory pool limit: `FairSpillPool::new(4 * 1024 * 1024 * 1024)`

SQL:
`SELECT * FROM rnd rnd1 JOIN rnd rnd2 ON rnd1."s3_drive" = rnd2."s3_drive"`

**Expected behavior**

It should return`ResourcesExhausted` error with configured `MemoryPool`

**Additional context**

A part of this was described in the discussion here: https://github.com/apache/arrow-datafusion/issues/5108#issuecomment-1412491794, but there was mentioned the regression. 
This example isn't regression and it's reproducible for old versions

### PR DESCRIPTION
**Pull request by korowa**
# Which issue does this PR close?

Closes #5162.
Part of #5220.

# Rationale for this change

Memory management for 1/4 Join operators

# What changes are included in this PR?

 - `CrossJoinExec.reservation` -- the actual reservation, used in `load_left_input` closure to reserve memory for build-side data, and in stream polling function to free memory after probe-side stream has been exhausted. Another point where memory gets freed is a destructor of reservation (i.e. in case when parent operator has `LIMIT`), so it's not required to store reservation as a stream attribute, but explicitly calling `free` seems like more correct way
 - `BuildProbeJoinMetrics` -- at this moment CrossJoinExec lacks metrics which could be exposed in `explain analyze` -- the counters from `CrossJoinStream` has been moved to `BuildProbeJoinMetrics` structure -- it could also be used in followups for both `NestedLoopJoinExec` and `HashJoinExec` as they all share almost the same "build & probe" idea, so their metrics could be similar.

# Are these changes tested?

Test cases for normal cross join execution and asserting error in case of overallocation attempt have been added

# Are there any user-facing changes?

Not really, but should help to avoid OOM and fail with `Resources exhausted` error.

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @korowa  -- I plan to review this carefully tomorrow

File: datafusion/core/src/physical_plan/joins/cross_join.rs [hunk 1]
```diff
@@ -434,15 +453,7 @@ impl CrossJoinStream {
                     Some(result)
                 }
                 other => {
-                    debug!(
```
Comment (alamb) [inline datafusion/core/src/physical_plan/joins/cross_join.rs:437]: I think it is a nice improvement that the metrics are now included in things like `EXPLAIN ANALYZE`

**Comment by korowa** [comment]
Thank you for the review @alamb!

> I think we should also add sql level test for the memory limit code in; I think it should be easy to add one to https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/memory_limit.rs. I am happy to add such a test if you prefer as a follow on PR.

Seems like new `test_overallocation` is sort of duplicate of `memory_limit` tests  -- so, I guess, it'll be better to add new `memory_limit` test instead of existing one in this PR

**Review by alamb — APPROVED** [review]
Thanks @korowa

**Pull request merged**